http-clients/src/main/java/org/triplea/http/client/lobby/chat/LobbyChatClient.java

```diff
@@ -1,10 +1,98 @@
 package org.triplea.http.client.lobby.chat;
 
-import lombok.experimental.UtilityClass;
+import com.google.common.annotations.VisibleForTesting;
+import java.net.URI;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.function.Consumer;
+import lombok.extern.java.Log;
+import org.triplea.domain.data.ApiKey;
+import org.triplea.domain.data.PlayerName;
+import org.triplea.http.client.lobby.chat.events.client.ClientMessageEnvelope;
+import org.triplea.http.client.lobby.chat.events.client.ClientMessageFactory;
+import org.triplea.http.client.lobby.chat.events.server.ChatMessage;
+import org.triplea.http.client.lobby.chat.events.server.PlayerSlapped;
+import org.triplea.http.client.lobby.chat.events.server.ServerMessageEnvelope;
+import org.triplea.http.client.lobby.chat.events.server.StatusUpdate;
+import org.triplea.http.client.web.socket.GenericWebSocketClient;
 
 /** Core websocket client to communicate with lobby chat API. */
-@UtilityClass // TODO: Project#12 remove utility class annotation
+@Log
 public class LobbyChatClient {
   public static final String WEBSOCKET_PATH = "/lobby/chat/websocket";
-  // TODO: Project#12 add implementation
+
+  private final GenericWebSocketClient<ServerMessageEnvelope, ClientMessageEnvelope>
+      webSocketClient;
+  private final ClientMessageFactory outboundMessageFactory;
+  private final InboundChat inboundChat;
+
+  public LobbyChatClient(final URI lobbyUri, final ApiKey apiKey) {
+    this(new InboundChat(URI.create(lobbyUri + WEBSOCKET_PATH)), new ClientMessageFactory(apiKey));
+  }
+
+  @VisibleForTesting
+  LobbyChatClient(final InboundChat inboundChat, final ClientMessageFactory clientEventFactory) {
+    this.inboundChat = inboundChat;
+    webSocketClient = inboundChat.getWebSocketClient();
+    outboundMessageFactory = clientEventFactory;
+  }
+
+  public static LobbyChatClient newClient(final URI lobbyUri, final ApiKey apiKey) {
+    return new LobbyChatClient(lobbyUri, apiKey);
+  }
+
+  public void slapPlayer(final PlayerName playerName) {
+    webSocketClient.send(outboundMessageFactory.slapMessage(playerName));
+  }
+
+  public void sendChatMessage(final String message) {
+    webSocketClient.send(outboundMessageFactory.sendMessage(message));
+  }
+
+  public void close() {
+    webSocketClient.close();
+  }
+
+  public Collection<ChatParticipant> connect() {
+    webSocketClient.send(outboundMessageFactory.connectToChat());
```

Comment:
Ok, this confused me for a minute:
`connectToChat()` sounded to me like this method connects to the chat and then this method sends some sort of acknowledgement packet to the server which I found weird.
Looking the code up it makes complete sense now, but you might want to consider renaming the method in the future.

Reply:
I'm not sure of any renaming that would help.
`connect` for `lobbyChatClient` we are stuck with because of how the previous chat worked. The (API) users of the chat transmitter interface are more aware than they should be of async connect vs sync connect, but ultimately we are trying to hide that detail.

For the message factory, a 'connect' message is an explicit connection to chat, logical. Until that is done, chat messages are not routed back. There is also a difference between connecting to the server, and actually connecting to the chat. At some point as well blocking needs to be added to rejected the banned users.

Let me know if you can come up with any ideas for better naming, I think at trouble here is just the multiple different API perspectives

Reply:
@DanVanAtta Ah, I wasn't clear enough:
What I was trying to say is that I didn't realise `connectToChat()` just builds a message to send, and doesn't immediately send it as the verb 'connect' implies.

I mean obviously the name of the variable gives it away that the class is just a factory, but confused me for a while